client/clipboards/Wayland/src/wayland.c

```diff
@@ -0,0 +1,528 @@
+    .capabilities = seat_capabilities_handler,
+    .name = seat_name_handler
+};
+
+// Registry-handling listeners.
+
+static void registry_global_handler(void * data,
+    struct wl_registry * wl_registry, uint32_t name, const char * interface,
+    uint32_t version) {
+  if (!strcmp(interface, wl_data_device_manager_interface.name))
+    this->data_device_manager = wl_registry_bind(this->registry, name,
+        &wl_data_device_manager_interface, 3);
+  else if (!strcmp(interface, wl_seat_interface.name) && !this->seat)
+    // TODO: multi-seat support.
+    this->seat = wl_registry_bind(this->registry, name,
+        &wl_seat_interface, 1);
+}
+
+static void registry_global_remove_handler(void * data,
+    struct wl_registry * wl_registry, uint32_t name) {
+  // Do nothing.
+}
+
+static const struct wl_registry_listener registry_listener = {
+  .global = registry_global_handler,
+  .global_remove = registry_global_remove_handler,
+};
+
+// Destination client handlers.
+
+static void wl_data_handle_offer(void * data, struct wl_data_offer * offer,
+    const char * mimetype)
+{
+  enum LG_ClipboardData type = mimetype_to_cb_type(mimetype);
+  // Oftentimes we'll get text/html alongside text/png, but would prefer to send
+  // image/png. In general, prefer images over text content.
+  if (type != LG_CLIPBOARD_DATA_NONE &&
+      (this->stashed_type == LG_CLIPBOARD_DATA_NONE ||
+       this->stashed_type == LG_CLIPBOARD_DATA_TEXT))
+  {
+    this->stashed_type = type;
+    if (this->stashed_mimetype)
+      free(this->stashed_mimetype);
+    this->stashed_mimetype = strdup(mimetype);
+  }
+}
+
+static const struct wl_data_offer_listener data_offer_listener = {
+  .offer = wl_data_handle_offer,
+};
+
+static void wl_data_device_handle_data_offer(void * data,
+    struct wl_data_device * data_device, struct wl_data_offer * offer)
+{
+  wl_data_offer_add_listener(offer, &data_offer_listener, NULL);
+}
+
+static void wl_data_device_handle_selection(void * data,
+    struct wl_data_device * data_device, struct wl_data_offer * offer)
+{
+  if (this->stashed_type == LG_CLIPBOARD_DATA_NONE || !offer)
+    return;
+
+  int fds[2];
+  if (pipe(fds) < 0)
+  {
+    perror("pipe");
+    abort();
+  }
+
+  wl_data_offer_receive(offer, this->stashed_mimetype, fds[1]);
+  close(fds[1]);
+
+  wl_display_roundtrip(this->display);
+
+  if (this->stashed_contents)
+  {
+    free(this->stashed_contents);
+    this->stashed_contents = NULL;
+  }
+
+  size_t size = 4096, num_read = 0;
+  uint8_t * buf = (uint8_t *) malloc(size);
+  while (true)
+  {
+    ssize_t result = read(fds[0], buf + num_read, size - num_read);
+    if (result < 0)
+    {
+      perror("read");
+      abort();
+    }
+
+    if (result == 0)
+    {
+      buf[num_read] = 0;
+      break;
+    }
+
+    num_read += result;
+    if (num_read >= size)
+    {
+      size *= 2;
+      void * nbuf = realloc(buf, size);
+      if (!nbuf) {
+        perror("realloc");
```

Comment:
please don't use `perror`, or `abort`, use `DEBUG_ERROR` and implement a graceful exit.